src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs

```diff
@@ -43,16 +44,38 @@ public HttpsConnectionMiddleware(ConnectionDelegate next, HttpsConnectionAdapter
                 throw new ArgumentNullException(nameof(options));
             }
 
+            _options = options;
+            _logger = loggerFactory.CreateLogger<HttpsConnectionMiddleware>();
+
             // This configuration will always fail per-request, preemptively fail it here. See HttpConnection.SelectProtocol().
             if (options.HttpProtocols == HttpProtocols.Http2)
             {
                 if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
                 {
                     throw new NotSupportedException(CoreStrings.HTTP2NoTlsOsx);
                 }
-                else if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && Environment.OSVersion.Version < new Version(6, 2))
+                else if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
                 {
-                    throw new NotSupportedException(CoreStrings.HTTP2NoTlsWin7);
+                    var enableHttp2OnWindows81 = AppContext.TryGetSwitch(EnableWindows81Http2, out var enabled) && enabled;
+                    if (Environment.OSVersion.Version < new Version(6, 3)
+                        || (Environment.OSVersion.Version < new Version(10, 0) && !enableHttp2OnWindows81))
+                    {
+                        throw new NotSupportedException(CoreStrings.HTTP2NoTlsWin81);
+                    }
+                }
+            }
+
+            if (options.HttpProtocols == HttpProtocols.Http1AndHttp2)
+            {
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                {
+                    var enableHttp2OnWindows81 = AppContext.TryGetSwitch(EnableWindows81Http2, out var enabled) && enabled;
+                    if (Environment.OSVersion.Version < new Version(6, 3)
+                        || (Environment.OSVersion.Version < new Version(10, 0) && !enableHttp2OnWindows81))
```

Comment:
This logic is duplicated (line 59 and line 72). Split into its own method and reuse.